File: bids_make_spm_multi_regression_files.m
function bids_make_spm_multi_regression_files( sub, experiment, tasks, new_format, overwrite )
%BIDS_MAKE_SPM_MULTI_REGRESSION_FILE 
%   convert FMRIPREP confounds.tsv files to spm multiple conditions format
%   regressors: 6 motion regressors, aCompCor (6 comps), and FD
%   (framewise-displacement). 
%   component choices as recommended by Chris Gorkolewski on Neurostars
%   (https://neurostars.org/t/confounds-from-fmriprep-which-one-would-you-use-for-glm/326)
%
%   Author: Nicholas Blauch
%   Updated 4/19/19 to reflect new FMRIPREP naming system
%   Updated 8/7/18 to avoid rewriting pre-existing files
%   Updated 4/30/18 to take sub as an integer, and for support with
%   off-server computation using get_bids_dir

bids_dir = get_bids_dir(experiment);

% default to old format for backwards compatibility
if nargin < 4
    new_format = 0;
end
if nargin < 5
    overwrite = 0;
end

sub_deriv_dir = sprintf('%s/derivatives/fmriprep/sub-%02d',bids_dir,sub);
out_dir =sprintf('%s/derivatives/matlab/spm/multiple_regressors/sub-%02d',bids_dir,sub);
if ~exist(out_dir,'dir')
    mkdir(out_dir)
end
multisess_format = ~(~exist([sub_deriv_dir,'/ses-01'],'dir'));
n_sess = 0;
if multisess_format
    ii = 1;
    while exist(sprintf('%s/ses-%02d',sub_deriv_dir,ii),'dir')
        n_sess = ii;
        ii = ii + 1;
    end
end

if multisess_format
    for sess = 1:n_sess
        sess_dir = sprintf('%s/ses-%02d',sub_deriv_dir,sess);
        if ~exist(sprintf('%s/ses-%02d',out_dir,sess),'dir')
            mkdir(sprintf('%s/ses-%02d',out_dir,sess))
        end
        for task = 1:length(tasks)
            exp_run = 0;
            if new_format
                name = sprintf('%s/func/sub-%02d_ses-%02d_task-%s_run-%02d_desc-confounds_regressors.tsv',sess_dir,sub,sess,tasks{task},exp_run+1);
            else
                name = sprintf('%s/func/sub-%02d_ses-%02d_task-%s_run-%02d_bold_confounds.tsv',sess_dir,sub,sess,tasks{task},exp_run+1);
            end
            while exist(name,'file')
                exp_run = exp_run + 1;

                fname = sprintf('%s/ses-%02d/sub-%02d_ses-%02d_task-%s_run-%02d_multiregressors.txt',out_dir,sess,sub,sess,tasks{task},exp_run);
                if exist(fname,'file')
                    name = strrep(name, sprintf('run-%02d', exp_run), sprintf('run-%02d', exp_run+1));
                    continue
                end
                
                confounds = tdfread(name);
                [translation, regressors] = analyze_confounds(confounds, new_format);
                fprintf('\n sub-%02d_ses-%02d_task-%s_run-%02d: max displacement from start: %1.3f mm \n',sub,sess,tasks{task}, exp_run, max(translation(1,:)));
                %change NAN to regressor mean for regressors with leading or other NAN.
                for regressor = 1:size(regressors,2)
                    regressors(isnan(regressors(:,regressor))',regressor) = mean(regressors(~isnan(regressors(:,regressor)),regressor));
                end
                dlmwrite(fname,regressors)
                name = strrep(name, sprintf('run-%02d', exp_run), sprintf('run-%02d', exp_run+1));
            end
        end
    end
    
else
    for task = 1:length(tasks)
        exp_run = 0;
        if new_format
            name = sprintf('%s/func/sub-%02d_task-%s_run-%02d_desc-confounds_regressors.tsv',sub_deriv_dir,sub,tasks{task},exp_run+1);
        else
            name = sprintf('%s/func/sub-%02d_task-%s_run-%02d_bold_confounds.tsv',sub_deriv_dir,sub,tasks{task},exp_run+1);
        end
        while exist(name,'file')
            exp_run = exp_run + 1;
            
            fname = sprintf('%s/sub-%02d_task-%s_run-%02d_multiregressors.txt',out_dir,sub,tasks{task},exp_run);
            if ~overwrite
                if exist(fname,'file')
                    name = strrep(name, sprintf('run-%02d', exp_run), sprintf('run-%02d', exp_run+1));
                    continue
                end
            end
            
            confounds = tdfread(name);
            [translation, regressors] = analyze_confounds(confounds, new_format);
            fprintf('\n sub-%02d_task-%s_run-%02d: max displacement from start: %1.3f mm \n',sub,tasks{task}, exp_run, max(translation(1,:)));
            %change NAN to regressor mean for regressors with leading or other NAN.
            for regressor = 1:size(regressors,2)
                regressors(isnan(regressors(:,regressor))',regressor) = mean(regressors(~isnan(regressors(:,regressor)),regressor));
            end
            dlmwrite(fname,regressors);
            
            name = strrep(name, sprintf('run-%02d', exp_run), sprintf('run-%02d', exp_run+1));
            
        end
    end
end

end

function [translation, regressors] = analyze_confounds(confounds, new_format)
    if new_format
        for ii = 1:length(confounds.t_comp_cor_00)
            confounds.FD_double(ii) = str2double(confounds.framewise_displacement(ii,:)); 
        end
        translation = dist([confounds.trans_x,confounds.trans_y,confounds.trans_z]');
        regressors = [confounds.a_comp_cor_00,confounds.a_comp_cor_01,confounds.a_comp_cor_02,confounds.a_comp_cor_03,...
            confounds.a_comp_cor_04,confounds.a_comp_cor_05,confounds.trans_x,confounds.trans_y,confounds.trans_z,confounds.rot_x,...
            confounds.rot_y,confounds.rot_z,confounds.FD_double'];                    
    else
        for ii = 1:length(confounds.tCompCor00)
            confounds.FD_double(ii) = str2double(confounds.FramewiseDisplacement(ii,:)); 
        end
        translation = dist([confounds.X,confounds.Y,confounds.Z]');
        regressors = [confounds.aCompCor00,confounds.aCompCor01,confounds.aCompCor02,confounds.aCompCor03,...
            confounds.aCompCor04,confounds.aCompCor05,confounds.X,confounds.Y,confounds.Z,confounds.RotX,...
            confounds.RotY,confounds.RotZ,confounds.FD_double'];
    end
end

File: bids_make_spm_multi_regression_files.py
# rewrite of bids_make_spm_multi_regression_files.m for python, ignoring the old format, and extending to allow for one-hot motion regression, and % variance-driven selection of number of CompCor components. does not support multiple session format currently.

import os
import sys
import numpy as np
import pandas as pd
import argparse
import glob
import pickle
import json
from scipy.spatial.distance import pdist, squareform

def analyze_confounds(confounds, confounds_json, compcorvar=None, compcornum=6, fdthresh=None, dvarthresh=None, frames=1, autooutlier=True):
    assert compcorvar is not None or compcornum is not None and not (compcorvar is not None and compcornum is not None), 'Must specify either compcorvar or compcornum, but not both'
    if compcorvar is not None:
        assert compcorvar <= .5, 'compcorvar must be less than or equal to .5'
        thresh_reached = False
        ii = 0
        while not thresh_reached:
            cumvar = confounds_json[f'a_comp_cor_{ii:02d}']['CumulativeVarianceExplained']
            if cumvar > compcorvar:
                thresh_reached = True
                compcornum = ii+1
            else:
                ii += 1
    translation = pdist(np.stack((confounds.trans_x, confounds.trans_y, confounds.trans_z),1))
    max_vals = {'max_translation':np.max(squareform(translation)[0,:]), 
                'max_FD': np.max(confounds.framewise_displacement), 
                'max_DVARS': np.max(confounds.std_dvars),
                'mean_DVARS': np.mean(confounds.std_dvars),
                'compcornum': compcornum,
                }
    all_confounds = [f'a_comp_cor_{ii:02d}' for ii in range(compcornum)]
    all_confounds += [f'trans_{ii}' for ii in ['x', 'y', 'z']]
    all_confounds += [f'rot_{ii}' for ii in ['x', 'y', 'z']]
    all_confounds += ['framewise_displacement']
    regressors = confounds[all_confounds].to_numpy()
    if fdthresh is not None or dvarthresh is not None:
        assert not autooutlier, 'Cannot specify both fdthresh/dvarthresh and autooutlier'
        fdthresh = np.inf if fdthresh is None else fdthresh
        dvarthresh = np.inf if dvarthresh is None else dvarthresh
        bad_tps = np.argwhere(
            np.logical_or(
                confounds.framewise_displacement.to_numpy() > fdthresh,
                confounds.std_dvars.to_numpy() > dvarthresh,
            )).flatten()
        exclude_tps = []
        if len(bad_tps) > 0:
            for tp in bad_tps:
                exclude_tps += list(range(tp-frames, tp+frames+1))
            exclude_tps = np.unique(exclude_tps)
            onehot_exclude_tps = np.eye(confounds['dvars'].shape[0])[:,exclude_tps]
            regressors = np.hstack((regressors, onehot_exclude_tps))
        max_vals['n_motion_outliers'] = len(exclude_tps)
    elif autooutlier:
        cols = [col for col in confounds.columns if 'motion_outlier' in col]
        regressors = np.hstack((regressors, confounds[cols].to_numpy()))
        max_vals['n_motion_outliers'] = len(cols)
    max_vals['n_confounds'] = regressors.shape[1]
    return pd.DataFrame(regressors), max_vals

def main(args):
    # compute backwards-compatible string tag for testing multiple versions of correction
    tag = ''
    if args.compcorvar is not None:
        tag += f'_ccvar-{args.compcorvar}'
    if args.compcornum != 6 and args.compcornum is not None:
        tag += f'_ccnum-{args.compcornum}'
    if args.fdthresh is not None:
        tag += f'_fdthresh-{args.fdthresh}'
    if args.autooutlier:
        tag += '_autooutlier'

    bids_dir = os.path.join(os.environ['BIDS'], args.experiment)
    sub_deriv_dir = os.path.join(bids_dir, 'derivatives', 'fmriprep', f'sub-{args.sub:02d}')
    out_dir = os.path.join(bids_dir, 'derivatives', 'matlab', 'spm', 'multiple_regressors', f'sub-{args.sub:02d}')
    for ii, task in enumerate(args.tasks):
        num_runs = len(glob.glob(f'{sub_deriv_dir}/func/*task-{task}*confounds_regressors.tsv'))
        for exp_run in range(1,num_runs+1):
            in_file = os.path.join(sub_deriv_dir, 'func', f'sub-{args.sub:02d}_task-{task}_run-{exp_run:02d}_desc-confounds_regressors.tsv')
            out_file = os.path.join(out_dir, f'sub-{args.sub:02d}_task-{task}_run-{exp_run:02d}{tag}_multiregressors.txt')
            if not os.path.exists(out_file) or args.overwrite:
                confounds = pd.read_csv(in_file, sep='\t')
                confounds_json = json.load(open(in_file.replace('.tsv', '.json')))
                regressors, max_vals = analyze_confounds(confounds, confounds_json,
                                                             compcorvar=args.compcorvar, compcornum=args.compcornum, fdthresh=args.fdthresh, frames=args.frames, autooutlier=args.autooutlier)
                regressors.to_csv(out_file, sep=',', index=False, header=False, float_format='%.7f')
                with open(out_file.replace('.txt', '_info.json'), 'w') as f:
                    json.dump(max_vals, f)
                print(f'wrote {out_file}')
                print(f'sub-{args.sub:02d}\ntask-{task}\nrun-{exp_run:02d}')
                for name, val in max_vals.items():
                    print(f'\t{name}: {val}')
            else:
                print(f'File {out_file} already exists, skipping')

parser = argparse.ArgumentParser(description='Generate SPM multi-regression files for BIDS dataset')
parser.add_argument('--experiment', type=str)
parser.add_argument('--sub', type=int)
parser.add_argument('--tasks', nargs='*', type=str)
parser.add_argument('--overwrite', action='store_true')
parser.add_argument('--compcorvar', type=float, default=None)
parser.add_argument('--compcornum', type=int, default=None)
parser.add_argument('--fdthresh', type=float, default=None, help='threshold for identifying outlier frames based on framewise displacement')
parser.add_argument('--dvarthresh', type=float, default=None, help='threshold for identifying outlier frames based on DVARS')
parser.add_argument('--frames', type=int, default=1, help='number of frames to exclude on each side of outlier frames')
parser.add_argument('--autooutlier', action='store_true', help='automatically identify outlier frames based FMRIPREP defaults of FD > 0.5, DVARS>1.5, frames=0')
args = parser.parse_args()

main(args)